presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java

```diff
@@ -138,51 +148,117 @@ public void stop()
         stopped = true;
     }
 
-    // This method may only be called if outstandingTasks == 0, or recursively after incrementing outstandingTasks
-    private void startLoadSplits()
+    private class HiveSplitLoaderTask
+            implements ResumableTask
     {
-        if (stopped || (fileIterators.isEmpty() && partitions.isEmpty())) {
-            return;
-        }
-        if (outstandingTasks.incrementAndGet() > maxPartitionBatchSize) {
-            outstandingTasks.decrementAndGet();
-            return;
-        }
-        executor.execute(() -> {
-            try {
-                loadSplits();
-                if (!hiveSplitSource.isQueueFull()) {
-                    // Start another task to replace this one
-                    startLoadSplits();
-                    // Ramp up if we're below the limit and the queue still isn't filled
-                    if (outstandingTasks.get() < maxPartitionBatchSize) {
-                        startLoadSplits();
-                    }
+        @Override
+        public TaskStatus process()
+        {
+            while (true) {
+                if (stopped) {
+                    return TaskStatus.finished();
                 }
-                if (outstandingTasks.decrementAndGet() == 0) {
-                    // Only one thread will reach this point, and all other threads are guaranteed to have finished their work,
-                    // so it's safe to check the queues without further synchronization
-                    if (fileIterators.isEmpty() && partitions.isEmpty()) {
-                        hiveSplitSource.finished();
+                try {
+                    LoadSplitsStatus status;
+                    threeQueuesRwLock.readLock().lock();
+                    try {
+                        status = loadSplits();
                     }
+                    finally {
+                        threeQueuesRwLock.readLock().unlock();
+                    }
+                    if (status.isDidNotFindWork()) {
+                        invokeFinishedIfNecessary();
+                    }
+                    else if (status.isBlocked()) {
+                        return TaskStatus.continueOn(status.getFuture());
+                    }
+                }
+                catch (Exception e) {
+                    hiveSplitSource.fail(e);
+                }
+            }
+        }
+    }
+
+    private void invokeFinishedIfNecessary()
+    {
+        if (partitions.isEmpty() && fileIterators.isEmpty()) {
```

Comment:
The return type of `loadSplits` has been simplified and the check is no longer duplicate.